dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitLabSourceCodeRepo.java

```diff
@@ -0,0 +1,377 @@
+
+            JsonElement jsonElement = new JsonParser().parse(projectJson);
+            if (jsonElement instanceof JsonArray) {
+                JsonArray jsonArray = jsonElement.getAsJsonArray();
+                for (JsonElement project : jsonArray) {
+                    JsonObject projectObject = project.getAsJsonObject();
+
+                    // What if username != namespace?
+                    if (projectObject.get("path_with_namespace").getAsString().equals(gitUsername + "/" + gitRepository)) {
+                        id = projectObject.get("id").getAsString();
+                        if (reference == null) {
+                            branch = projectObject.get("default_branch").getAsString();
+                        }
+                        break;
+                    }
+                }
+            }
+
+        }
+
+        // Get file contents
+        if (id != null || branch != null) {
+            return getFileContentsFromId(id, branch, fileName);
+        }
+
+        return content;
+    }
+
+    @Override public String getOrganizationEmail() {
+        return null;
+    }
+
+    @Override public Map<String, String> getWorkflowGitUrl2RepositoryId() {
+        Map<String, String> reposByGitUrl = new HashMap<>();
+        String projectsUrl = GITLAB_API_URL + "projects";
+
+        Optional<String> asString = ResourceUtilities.asString(projectsUrl, gitlabTokenContent, client);
+
+        if (asString.isPresent()) {
+            String projectJson = asString.get();
+
+            JsonElement jsonElement = new JsonParser().parse(projectJson);
+            if (jsonElement instanceof JsonArray) {
+                JsonArray jsonArray = jsonElement.getAsJsonArray();
+                for (JsonElement project : jsonArray) {
+                    JsonObject projectObject = project.getAsJsonObject();
+                    String gitlabUrl = projectObject.get("ssh_url_to_repo").getAsString();
+                    String id = projectObject.get("path_with_namespace").getAsString();
+                    reposByGitUrl.put(gitlabUrl, id);
+                }
+            }
+        }
+
+        return reposByGitUrl;
+    }
+
+    @Override public Workflow initializeWorkflow(String repositoryId) {
+        Workflow workflow = new Workflow();
+
+        // Does this split not work if name has a slash?
+        String[] id = repositoryId.split("/");
+        String owner = id[0];
+        String name = id[1];
+
+        // Setup workflow
+        workflow.setOrganization(owner);
+        workflow.setRepository(name);
+
+        final String gitUrl = GITLAB_GIT_URL_PREFIX + repositoryId + GITLAB_GIT_URL_SUFFIX;
+        workflow.setGitUrl(gitUrl);
+        workflow.setLastUpdated(new Date());
+        workflow.setPath(workflow.getPath());
+
+        return workflow;
+    }
+
+    @Override public Workflow setupWorkflowVersions(String repositoryId, Workflow workflow, Optional<Workflow> existingWorkflow,
```

Comment:
setupWorkflowVersions in particular looks very repetitive with https://github.com/ga4gh/dockstore/blob/develop/dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java and https://github.com/ga4gh/dockstore/blob/develop/dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/BitBucketSourceCodeRepo.java